### PR TITLE
github workflow to automate approving and merging PRs

### PR DESCRIPTION
**Pull request by tnsardesai**
**jira**: Passion week 

**Overview:**

- I added hmarr/auto-approve-action@5d04a5ca6da9aeb8ca9f31a5239b96fc3e003029 to be allowed at org level.
- Tested this with https://github.com/Clever/dependency-failure-diagram-generator/blob/master/.github/workflows/dependabot.yml workflow which is pretty straight forward
- Wrote a github card to explain how to set this up https://app.getguru.com/card/cbjEdoAi/Dependabot-and-auto-merging
- See me testing multiple different scenarios here! https://github.com/Clever/dependency-failure-diagram-generator/pull/219
  - PR says approved and merged by `github-actions`
- Coincidentally here is an example of when this doesn't work https://github.com/Clever/dependency-failure-diagram-generator/pull/220. The previous PR was merged after this was opened leading to merge conflicts. In reality this should be rare. If it is a big problem then we can use https://github.com/pascalgn/automerge-action instead of the `gh` cli
  - Even this case tho you can see that PR is approved and auto merge is enabled, we just need to say `@dependabot rebase` and it should work. 
- Finally here is the automated deployment!!!! https://clever.slack.com/archives/C0345BFU137/p1645727815460149 🥳 🥳 🥳 

**Rollout:**
- https://github.com/Clever/dependency-failure-diagram-generator/blob/master/.github/workflows/dependabot.yml#L10 should point to `master`

### CODE REVIEW

File: .github/workflows/auto-approve-merge.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+on: workflow_call
+
+jobs:
+  auto-approve-merge-dependency-updates:
+    name: Approve and Enable auto merge on dependabot PRs
+    runs-on: ubuntu-latest
+    if: github.actor == 'dependabot[bot]'
+    steps:
+    # Use github cli to enable auto approve. Only works on repos which have auto merge enabled and branch protections setup which is currently blocking a merge
+    # for example still requires 1 review or a required ci step has not succedded.
```
Comment (nbhatia823) [inline .github/workflows/auto-approve-merge.yml:15]: typo: succeeded

**Comment by tnsardesai** [comment]
newest commit adds a workaround for approving PRs https://github.com/Clever/dependency-failure-diagram-generator/pull/222 by adding a label.

File: .github/workflows/auto-approve-merge.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    # Use github cli to enable auto approve. Only works on repos which have auto merge enabled and branch protections setup which is currently blocking a merge
+    # for example still requires 1 review or a required ci step has not succeeded.
+    - if: github.actor == 'dependabot[bot]'
+      name: Enable Automerge
+      run: gh pr merge --auto --merge "$PR_URL"
+      env:
+        PR_URL: ${{ github.event.pull_request.html_url }}
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    # v2.1.0 as of 02/23/2022, reviewed by eng-security.
+    - if: github.actor == 'dependabot[bot]' || contains( github.event.pull_request.labels.*.name, 'auto approve')
```
Comment (taylor-sutton) [inline .github/workflows/auto-approve-merge.yml:23]: Is this safe to use on public repos? For example, do non-Clever folks have access to add labels?

Comment (taylor-sutton) [reply .github/workflows/auto-approve-merge.yml:23]: and/or can a malicious actor impersonate Dependabot.

Comment (tnsardesai) [reply .github/workflows/auto-approve-merge.yml:23]: - non-Clever folks should not have access to add labels. You need to have at least "triage" level permissions to apply labels but non-clever folks should just have read permissions.

- Not sure about if someone can impersonate the github actor. I can investigate more on this. Maybe it makes sense to replace it with `github.event.pull_request.user.login`

Comment (tnsardesai) [reply .github/workflows/auto-approve-merge.yml:23]: `github.actor` is user that initiated the workflow run whereas `github.event.pull_request.user.login` is user that created the pull request

Comment (tnsardesai) [reply .github/workflows/auto-approve-merge.yml:23]: I am reading https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections and looks like `github.actor` can be trusted. It is stuff like `github.event.issue.title` or `github.event.pull_request.body` where we need to be careful as that can be edited by the any user.

**Review by taylor-sutton — APPROVED** [review]
LGTM but one last question: is this the right directory to put this in inside ci-scripts? i.e. we might want to reserve this directory for things that apply to this repo only, rather than shared scripts.

**Comment by tnsardesai** [comment]
> LGTM but one last question: is this the right directory to put this in inside ci-scripts? i.e. we might want to reserve this directory for things that apply to this repo only, rather than shared scripts.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow unfortunately it is a requirement from github that this has to live in `.github/workflows` and we cannot create sub directories either.

Maybe I can rename this file to be `reusable-auto-approve-merge.yml` and we can build a convention of using `reusable` prefix with reusable workflows and any workflows that apply to this repo can be named whatever.

**Comment by taylor-sutton** [comment]
> > LGTM but one last question: is this the right directory to put this in inside ci-scripts? i.e. we might want to reserve this directory for things that apply to this repo only, rather than shared scripts.
> 
> https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow unfortunately it is a requirement from github that this has to live in `.github/workflows` and we cannot create sub directories either.
> 
> Maybe I can rename this file to be `reusable-auto-approve-merge.yml` and we can build a convention of using `reusable` prefix with reusable workflows and any workflows that apply to this repo can be named whatever.

👍

**Pull request merged**